conmon-rs/server/src/child_reaper.rs

```diff
@@ -1,169 +1,148 @@
+impl ChildReaper {
+    pub fn get(&self, id: String) -> Result<ReapableChild> {
+        let locked_grandchildren = Arc::clone(&self.grandchildren);
+        let lock = locked_grandchildren.lock().unwrap();
+        let r = lock.get(&id).context("")?.clone();
+        drop(lock);
+        Ok(r)
     }
 
-    pub fn create_child<P, I, S>(&self, cmd: P, args: I) -> Result<std::process::ExitStatus>
+    pub async fn create_child<P, I, S>(
+        &self,
+        cmd: P,
+        args: I,
+        console: Option<Console>,
+        pidfile: PathBuf,
+    ) -> Result<i32>
     where
         P: AsRef<std::ffi::OsStr>,
         I: IntoIterator<Item = S>,
         S: AsRef<std::ffi::OsStr>,
     {
-        let mut cmd = std::process::Command::new(cmd);
+        let mut cmd = tokio::process::Command::new(cmd);
         cmd.args(args);
-
-        // To prevent an error: "No child processes (os error 10)",
-        // we prevent our reaper thread from calling waitpid until this child has completed.
-        // Inspired by
-        // https://github.com/kata-containers/kata-containers/blob/828a304883a4/src/agent/src/signal.rs#L27
-        let wait_lock = Arc::clone(&self.wait_lock);
-        let _lock = wait_lock
-            .lock()
-            .map_err(|e| format_err!("lock waitlock: {}", e))?;
-
         cmd.spawn()
-            .map_err(|e| format_err!("spawn child process: {}", e))?
+            .context("spawn child process: {}")?
             .wait()
-            .map_err(|e| format_err!("wait for child process: {}", e))
+            .await?;
+
+        if let Some(console) = console {
+            let _ = console
+                .wait_connected()
+                .context("wait for console socket connection");
+        }
+
+        let grandchild_pid = tokio::fs::read_to_string(pidfile)
+            .await?
+            .parse::<i32>()
+            .context("grandchild pid parse error")?;
+
+        Ok(grandchild_pid)
     }
 
-    pub fn watch_grandchild(&self, child: &Child) -> Result<()> {
+    pub fn watch_grandchild(&self, child: Child) -> Result<()> {
         let locked_grandchildren = Arc::clone(&self.grandchildren);
         let mut map = locked_grandchildren
             .lock()
             .map_err(|e| format_err!("lock grandchildren: {}", e))?;
-
-        let reapable_grandchild = ReapableChild::from_child(child);
-        if let Some(old) = map.insert(child.pid, reapable_grandchild) {
-            bail!("Repeat PID for container {} found", old.id);
-        }
+        let reapable_grandchild = ReapableChild::from_child(&child);
+        let killed_channel = reapable_grandchild.watch();
+        map.insert(child.id, reapable_grandchild);
+        let cleanup_grandchildren = locked_grandchildren.clone();
+        let pid = child.pid;
+        tokio::task::spawn(async move {
+            killed_channel.await.expect("no error on channel");
+            if let Err(e) = Self::forget_grandchild(&cleanup_grandchildren, pid) {
```

Comment:
also, if we're waiting on the channel here to forget the granchild, why don't we just call waitpid within this task, rather than needing to sync between two tasks?

Reply:
since this is the 'sync' function I am thinking we want to wait on the async task to complete. Ideally, we want less code and usually the 'sync' version is just a block on the async version.